examples/imagenet/train.py

```diff
@@ -0,0 +1,726 @@
+    # by retrieving local rank from environment variable LOCAL_RANK.
+    if args.torchrun:
+        args.local_rank = int(os.environ["LOCAL_RANK"])
+        args.local_world_size = int(os.environ["LOCAL_WORLD_SIZE"])
+
+    args.distributed = args.multiprocessing_distributed or args.local_rank >= 0
+    ngpus_per_node = torch.cuda.device_count()
+
+    if args.multiprocessing_distributed:
+        # Use `torch.multiprocessing.spawn` to launch distributed processes: the
+        # main_worker process function.
+        mp.spawn(main_worker, nprocs=ngpus_per_node, args=(ngpus_per_node, args))
+    elif args.local_rank >= 0:
+        # Use `torch.distributed.launch` or `turchrun` or `slurm`.
+        # Simply call `main_worker` at `local_rank`.
+        main_worker(args.local_rank, ngpus_per_node, args)
+    else:
+        # Use a specific GPU.
+        # Simply call main_worker function.
+        main_worker(args.gpu, ngpus_per_node, args)
+
+
+def main_worker(gpu, ngpus_per_node, args):
+    args.gpu = gpu
+
+    if args.gpu is not None:
+        print(f"Use GPU {args.gpu} for training")
+
+    # DATA PARALLEL
+    if args.distributed:
+        if args.multiprocessing_distributed:
+            # Use `torch.multiprocessing`
+            # Spawn N processes, one for each GPU
+            dist.init_process_group(
+                backend=args.dist_backend,
+                init_method=args.dist_url,
+                world_size=ngpus_per_node,
+                rank=args.gpu,
+            )
+        else:
+            # Use `torchrun`, `torch.distributed.launch` or SLURM
+            # `MASTER_ADDR` and `MASTER_PORT` are already set as environment variables,
+            # so no need to pass to `init_process_group``.
+            dist.init_process_group(backend=args.dist_backend)
+
+        if args.local_world_size < 0:
+            args.local_world_size = dist.get_world_size()
+    else:
+        args.local_world_size = 1
+
+    # create model
+    if args.pretrained:
+        print("=> using pre-trained model '{}'".format(args.arch))
+        model = models.__dict__[args.arch](pretrained=True)
+    else:
+        print("=> creating model '{}'".format(args.arch))
+        model = models.__dict__[args.arch]()
+    if not torch.cuda.is_available():
+        print("using CPU, this will be slow")
+    elif args.distributed:
+        # DATA PARALLEL
+        # For multiprocessing distributed, DistributedDataParallel constructor
+        # should always set the single device scope, otherwise,
+        # DistributedDataParallel will use all available devices.
+        if args.gpu is not None:
+            torch.cuda.set_device(args.gpu)
+            model.cuda(args.gpu)
+            # NOTE: When using a single GPU per process and per
+            # DistributedDataParallel, we need to divide the batch size
+            # ourselves based on the total number of GPUs of the current node.
+            args.batch_size = int(args.batch_size / ngpus_per_node)
```

Comment:
I'm thinking about adding a `global_batch_size` and pass that to `ZeusDataLoader`, because of the rounding off here. WDYT? 
`global_batch_size` will be assigned to `self.train_batch_size`, as we have discussed, this variable is just for file naming.

Reply:
Since we'll be training with `int(batch_size / ngpus) * ngpus` samples, the global batch size isn't always `batch_size`.
The way frameworks like Megatron-LM deal with that issue is to statically check whether the requested batch size is divisible by the number of GPUs, and raise a `ValueError` if not.

Reply:
So we will rely on the user script to do the static check, right? I can put an additional NOTE.

Reply:
Yeah, and that `ZeusDataLoader` will just assume that the global batch size is going to be `batch_size * ngpus` or something.